packages/apputils/src/printing.ts

```diff
@@ -0,0 +1,177 @@
+  }
+
+  /**
+   * Returns whether an object implements a print method.
+   */
+  export function isPrintable(a: unknown): a is IPrintable {
+    if (typeof a !== 'object' || !a) {
+      return false;
+    }
+    return symbol in a;
+  }
+
+  /**
+   * Returns the print function for an object, or null if it does not provide a handler.
+   */
+  export function getPrintFunction(val: unknown): OptionalAsyncThunk {
+    if (isPrintable(val)) {
+      return val[symbol]();
+    }
+    return null;
+  }
+
+  /**
+   * Prints a widget by copying it's DOM node
+   * to a hidden iframe and printing that iframe.
+   */
+  export function printWidget(widget: Widget): Promise<void> {
+    return printContent(widget.node);
+  }
+
+  /**
+   * Prints a URL by loading it into an iframe.
+   *
+   * @param url URL to load into an iframe.
+   */
+  export async function printURL(url: string): Promise<void> {
+    const text = await (await ServerConnection.makeRequest(
+      url,
+      {},
+      ServerConnection.makeSettings()
+    )).text();
+    return printContent(text);
+  }
+
+  /**
+   * Prints a URL or an element in an iframe and then removes the iframe after printing.
+   */
+  async function printContent(textOrEl: string | HTMLElement): Promise<void> {
+    const isText = typeof textOrEl === 'string';
+    const iframe = createIFrame();
+
+    const parent = window.document.body;
+    parent.appendChild(iframe);
+    if (isText) {
+      iframe.srcdoc = textOrEl as string;
+      await resolveWhenLoaded(iframe);
+    } else {
+      iframe.src = 'about:blank';
+      setIFrameNode(iframe, textOrEl as HTMLElement);
+    }
+    const printed = resolveAfterEvent();
+    launchPrint(iframe.contentWindow);
+    // Once the print dialog has been dismissed, we regain event handling,
+    // and it should be safe to discard the hidden iframe.
+    await printed;
+    parent.removeChild(iframe);
+  }
+
+  /**
+   * Creates a new hidden iframe and appends it to the document
+   *
+   * Modified from
+   * https://github.com/joseluisq/printd/blob/eb7948d602583c055ab6dee3ee294b6a421da4b6/src/index.ts#L24
+   */
+  function createIFrame(): HTMLIFrameElement {
+    const el = window.document.createElement('iframe');
+
+    // We need both allow-modals and allow-same-origin to be able to
+    // call print in the iframe.
+    // We intentionally do not allow scripts:
+    // https://github.com/jupyterlab/jupyterlab/pull/5850#pullrequestreview-230899790
+    el.setAttribute('sandbox', 'allow-modals allow-same-origin');
+    const css =
+      'visibility:hidden;width:0;height:0;position:absolute;z-index:-9999;bottom:0;';
+    el.setAttribute('style', css);
+    el.setAttribute('width', '0');
+    el.setAttribute('height', '0');
+
+    return el;
+  }
+
+  /**
+   * Copies a node from the base document to the iframe.
+   */
+  function setIFrameNode(iframe: HTMLIFrameElement, node: HTMLElement) {
+    iframe.contentDocument.body.appendChild(node.cloneNode(true));
```

Comment:
Does this need to wait for the `load` event? My experience with the PDF viewer is that it was necessary, but maybe that's not the case here?

Reply:
I haven't found we needed to wait.